<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head><title> QoRTs: Quality of RNA-Seq Toolset </title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<link rel="stylesheet" type="text/css" href="../stylesheets/jar.css">
</head>
<body>
<div id="wrap">

<h1>QoRTs: Quality Of Rna-seq Tool Set</h1>

<blockquote>
  <p>Version 1.1.1 (Updated Mon Jun  6 14:57:02 EDT 2016)</p>

<p>(<a href="../index.html">back to main</a>) (<a href="index.html">back to java-utility help</a>)</p>
</blockquote>

<h2>Help for java command "makeJunctionTrack"</h2>

<h2>USAGE:</h2>

<pre><code>java [Java Options] -jar QoRTs.jar makeJunctionTrack [options] flatgff.gff.gz outfile
</code></pre>

<h2>DESCRIPTION:</h2>

<p>This utility takes the splice junction count files created by the QoRTs QC utility across multiple samples and creates a single merged splice junction 'bed' file that lists each splice junction along with the mean read-pair coverage counts (optionally, the mean normalized counts).This splice junction bed file can be used to visualize splice junction counts using the UCSC genome browser and other similar utilities. Note: Either the '--filenames' or the '--sampleList' option must be set! The sampleList option is generally used with the --infilePrefix and --infileSuffix options. Also note: This command only compiles the named splice junctions. For other unnamed splice junctions such as novel splice junctions with low coverage, or novel splice junctions that bridge multiple genes, use the makeAltJunctionTrack command instead.</p>

<h2>REQUIRED ARGUMENTS:</h2>

<h3>flatgff.gff.gz:</h3>

<blockquote>
  <p>The flattened gff file. (String)</p>
</blockquote>

<h3>outfile:</h3>

<blockquote>
  <p>The output bed file, or '-' to write to stdout. If the filename ends with ".gz" or ".zip" then the file will be compressed using the appropriate method. (String)</p>
</blockquote>

<h2>OPTIONAL ARGUMENTS:</h2>

<h3>--filenames file1,file2,file3,...:</h3>

<blockquote>
  <p>A comma-delimited list of wiggle files to merge. This is optional, and filenames can be inferred from --infilePrefix, --infileSuffix, and the --sampleList, if those options are specified. Either this option OR --sampleList MUST BE SPECIFIED. (CommaDelimitedListOfStrings)</p>
</blockquote>

<h3>--sampleList [sampleList.txt | - | samp1,samp2,samp3,...]:</h3>

<blockquote>
  <p>Either a comma-delimited list of sample id's or a file containing a list of sample id's. The file must either contain no title line, or contain a title line that includes a "sample.ID" column. Either this option OR --filenames MUST BE SPECIFIED. Note if the sample list is a file then it must end with the extension '.txt' (String)</p>
</blockquote>

<h3>--infilePrefix infilePrefix:</h3>

<blockquote>
  <p>A file prefix for all input junction count files. By default the full file path should be specified by the infile parameter. (String)</p>
</blockquote>

<h3>--infileSuffix infileSuffix:</h3>

<blockquote>
  <p>A file suffix for all input junction count files. By default the full file path should be specified by the infile parameter. (String)</p>
</blockquote>

<h3>--sizeFactorFile val:</h3>

<blockquote>
  <p>A file containing (at least) two columns: a list of sample ID's and their double-precision floating-point size factors. The first line must include at least two columns: "sample.ID" and "size.factor"If this option is set, all counts will be divided by the given normalization factors. The length must be the same as the length of infiles.If sample.ID's is not specified by the --sampleList or --sampleListFile parameters, then all listed samples will be merged. (String)</p>
</blockquote>

<h3>--sizeFactors val:</h3>

<blockquote>
  <p>A list of double-precision floating-point values. If this or any size factor option is set, all counts will be divided by the given normalization factors. The length must be the same as the number of files to merge. (CommaDelimitedListOfDoubles)</p>
</blockquote>

<h3>--calcMean:</h3>

<blockquote>
  <p>Flag to indicate that the splice junction counts should be averaged, rather than added up. (flag)</p>
</blockquote>

<h3>--stranded:</h3>

<blockquote>
  <p>Flag to indicate that data is stranded. (flag)</p>
</blockquote>

<h3>--rgb r,g,b:</h3>

<blockquote>
  <p>The rgb color for all the bed file lines. (String)</p>
</blockquote>

<h3>--trackTitle title:</h3>

<blockquote>
  <p>The title of the track. By default this will be the same as the title parameter. (String)</p>
</blockquote>

<h3>--additionalTrackOptions options:</h3>

<blockquote>
  <p>Additional track definition options, added to the track definition line. See the UCSC documentation for more information. (String)</p>
</blockquote>

<h3>--includeFullSpliceNames:</h3>

<blockquote>
  <p>Flag to indicate that full splice names, including gene ID, should be used. (flag)</p>
</blockquote>

<h3>--ignoreSizeFactors:</h3>

<blockquote>
  <p>Flag to indicate that this utility should ignore size factors even if they are found in the input listFile. (flag)</p>
</blockquote>

<h3>--nonflatgtf:</h3>

<blockquote>
  <p>Flag to indicate that instead of a "flattened" gff file, a standard-format gtf file has been specified. If this flag is raised, it will automatically create the standard flattened gtf information, in memory. It will not be written to disk (flag)</p>
</blockquote>

<h3>--skipAnnotatedJunctions:</h3>

<blockquote>
  <p>If this option is used, annotated splice junctions will not be included in the output file. Note: this only works if there are novel junctions in the input file. (flag)</p>
</blockquote>

<h3>--skipNovelJunctions:</h3>

<blockquote>
  <p>If this option is used, novel splice junctions will not be included in the output file. (flag)</p>
</blockquote>

<h3>--digits num:</h3>

<blockquote>
  <p>The number of digits after the decimal to include for counts. (Int)</p>
</blockquote>

<h3>--title title:</h3>

<blockquote>
  <p>A title to be prepended to each splice junction name (String)</p>
</blockquote>

<h3>--verbose:</h3>

<blockquote>
  <p>Flag to indicate that debugging information and extra progress information should be sent to stderr. (flag)</p>
</blockquote>

<h3>--quiet:</h3>

<blockquote>
  <p>Flag to indicate that only errors and warnings should be sent to stderr. (flag)</p>
</blockquote>

<h2>AUTHORS:</h2>

<p>Stephen W. Hartley, Ph.D. stephen.hartley (at nih dot gov)</p>

<h2>LEGAL:</h2>

<p>This software is "United States Government Work" under the terms of the United States Copyright  Act.  It was written as part of the authors' official duties for the United States Government and  thus cannot be copyrighted.  This software is freely available to the public for use without a  copyright notice.  Restrictions cannot be placed on its present or future use.  Although all reasonable efforts have been taken to ensure the accuracy and reliability of the  software and data, the National Human Genome Research Institute (NHGRI) and the U.S. Government  does not and cannot warrant the performance or results that may be obtained by using this software  or data.  NHGRI and the U.S. Government disclaims all warranties as to performance, merchantability  or fitness for any particular purpose.  In any work or product derived from this material, proper attribution of the authors as the source  of the software or data should be made, using "NHGRI Genome Technology Branch" as the citation.  NOTE: This package includes (internally) the sam-1.113.jar library from picard tools. That package uses the MIT license, which can be accessed using the command:  java -jar thisjarfile.jar help samjdkinfo</p>

</div></body></html>
