<h1>QoRTs: Quality Of Rna-seq Tool Set</h1>

<blockquote>
  <p>Version 1.1.0 (Updated Mon Jun  6 13:52:10 EDT 2016)</p>

<p>(<a href="../index.html">back to main</a>) (<a href="index.html">back to java-utility help</a>)</p>
</blockquote>

<h2>Help for java command "bamToWiggle"</h2>

<h2>USAGE:</h2>

<pre><code>java [Java Options] -jar QoRTs.jar bamToWiggle [options] infile.bam[,infile2.bam,...] trackTitlePrefix chromLengthFile outfilePrefix
</code></pre>

<h2>DESCRIPTION:</h2>

<p>Generates '.wig' wiggle files, suitable for use with the UCSC genome browser or similar tools. Wiggle files contain depth-of-coverage counts for all equally-sized windows across the entire genome. These depth-of-coverage counts can be either be by read (the default) or by read-pair.</p>

<h2>REQUIRED ARGUMENTS:</h2>

<h3>infile.bam[,infile2.bam,...]:</h3>

<blockquote>
  <p>The input sam or bam file or files, or '-' to read from stdin. Note: if you include more than one bam file, the list must be comma delimited with no whitespace! (CommaDelimitedListOfStrings)</p>
</blockquote>

<h3>trackTitlePrefix:</h3>

<blockquote>
  <p>The prefix for the title for the wiggle track. Track titles will be  (String)</p>
</blockquote>

<h3>chromLengthFile:</h3>

<blockquote>
  <p>The chrom length file. This should be a simple tab-delimited text file that includes each chromosome in the first column and the chromosome's length (in base-pairs) in the second column. If the wiggle file is intended for use with a standard genome on the UCSC genome browser, then the UCSC utility "fetchChromSizes" should be used to generate this file, as the UCSC genome browser and related utilities may have difficulties if the chrom lengths do not match the expected values. (String)</p>
</blockquote>

<h3>outfilePrefix:</h3>

<blockquote>
  <p>The output file prefix. If unstranded, this will produce one file named "outfilePrefix.unstranded.wig.gz". If stranded, this will produce two files: "outfilePrefix.fwd.wig.gz" and "outfilePrefix.rev.wig.gz". If the "--noGzipOutput" flag is raised then the output files will not have the ".gz" extension at the end. IMPORTANT NOTE: if the window size is set to any size other than the default of 100, the window size will be added to the filename. The filename will be  "outfilePrefix.win50.unstranded.wig.gz" for unstranded wiggles with a 50bp window, and so on. (String)</p>
</blockquote>

<h2>OPTIONAL ARGUMENTS:</h2>

<h3>--sizefactor float:</h3>

<blockquote>
  <p>The size factor, for normalization. Defaults to 1. If set, then the count of every cell will be divided by the given value. (Double)</p>
</blockquote>

<h3>--minMAPQ num:</h3>

<blockquote>
  <p>Filter reads for the given minimum MAPQ. Set to 0 to turn off mapq filtering. (Int)</p>
</blockquote>

<h3>--noGzipOutput:</h3>

<blockquote>
  <p>Flag to indicate that output should NOT be gzipped. (flag)</p>
</blockquote>

<h3>--negativeReverseStrand:</h3>

<blockquote>
  <p>Flag to indicate that the reverse strand should be counted in negative values. Can be useful for plotting stranded data on a single multiwig track, via trackhubs. (flag)</p>
</blockquote>

<h3>--countByReadPair:</h3>

<blockquote>
  <p>DEPRECIATED, nonfunctional. (flag)</p>
</blockquote>

<h3>--simpleCountByRead:</h3>

<blockquote>
  <p>Flag to indicate that this utility should count 'read coverage' rather than 'read-pair coverage'. If this flag is raised, then regions where a read-pair overlaps will be counted TWICE, one for each read. By default each read is counted separately, so a read-pair where the reads overlap will be double-counted. By default, overlapping regions of read-pairs are counted only once. (flag)</p>
</blockquote>

<h3>--testRun:</h3>

<blockquote>
  <p>Flag to indicate that only the first 100k reads should be read in, used for test runs. (flag)</p>
</blockquote>

<h3>--stranded:</h3>

<blockquote>
  <p>The stranded flag, used for strand-specific data. If this is set, then two wiggle files are created, one for each strand. (flag)</p>
</blockquote>

<h3>--stranded_fr_secondstrand:</h3>

<blockquote>
  <p>If this option is set, the data is assumed to be a fr_secondstrand type stranded library. By default the assumed library type for stranded data is fr_firststrand. (flag)</p>
</blockquote>

<h3>--singleEnded:</h3>

<blockquote>
  <p>Flag for single-end data. Note that many other options do not apply in this case (for example: option --countPairsTogether does nothing in single-end mode) (flag)</p>
</blockquote>

<h3>--keepMultiMapped:</h3>

<blockquote>
  <p>Flag to indicate that the tool should NOT filter out multi-mapped reads. Note that even with this flag raised this utility will still only use the 'primary' alignment location for each read. By default any reads that are marked as multi-mapped will be ignored entirely. (flag)</p>
</blockquote>

<h3>--includeTrackDefLine:</h3>

<blockquote>
  <p>DEPRECIATED, nonfunctional (flag)</p>
</blockquote>

<h3>--omitTrackDefLine:</h3>

<blockquote>
  <p>Flag to indicate that a track definition line should NOT be included. (flag)</p>
</blockquote>

<h3>--additionalTrackOptions "track options":</h3>

<blockquote>
  <p>More options for the track. For more information refer to the wiggle track definition on the UCSC genome browser website. (String)</p>
</blockquote>

<h3>--readGroup readGroupName:</h3>

<blockquote>
  <p>If this option is set, all analyses will be restricted to ONLY reads that are tagged with the given readGroupName (using an RG tag). This can be used if multiple read-groups have already been combined into a single bam file, but you want to summarize each read-group separately. (String)</p>
</blockquote>

<h3>--rgbColor rval,gval,bval:</h3>

<blockquote>
  <p>A comma delimited list of integers from 0-255, indicating the rgb color and alt-color for the wiggle file.This is equivalent to using --additionalTrackOptions "color=r,g,b altColor=r,g,b itemRgb=On" (String)</p>
</blockquote>

<h3>--noTruncate:</h3>

<blockquote>
  <p>The UCSC tool wigToBigWig only allows wiggle files in which every window is of equal size. This means that if the chromosome size is not divisible by the window size, a few bases are not counted on the end of the chromosome. Using this flag will cause this utility to NOT truncate off the final odd-sized window of each chromosome. However, be aware that this will mean that you CANNOT use the UCSC 'wigToBigWig' utility to convert the wiggle file to a (more efficient) 'bigWig' file. (flag)</p>
</blockquote>

<h3>--windowSize num:</h3>

<blockquote>
  <p>The size, in base-pairs, of each counting bin, or "window". Note: if this is set low the utility will take longer to run and will consume more memory. This utility is optimized for speed non memory use, and consumes (roughly) 8 bytes per window (plus maybe 250mb in overhead). The default window size is 100bp. (Int)</p>
</blockquote>

<h3>--nameSorted:</h3>

<blockquote>
  <p>Relevant for paired-end reads only. 
This flag is used to run QoRTs in "name-sorted" mode. This flag is optional, as under the default mode QoRTs will accept BAM files sorted by either name OR position. However, The only actual requirement in this mode is that read pairs be adjacent. 
Errors may occur if the SAM flags are inconsistent: for example, if orphaned reads appear with the "mate mapped" SAM flag set. (flag)</p>
</blockquote>

<h3>--coordSorted:</h3>

<blockquote>
  <p>DEPRECIATED: this mode is now subsumed by the default mode and as such this parameter is now nonfunctional.
Note that, in the default mode, for paired-end data QoRTs will accept EITHER coordinate-sorted OR name-sorted bam files. In "--nameSorted" mode, QoRTs ONLY accepts name-sorted bam files.
If a large fraction of the read-pairs are mapped to extremely distant loci (or to different chromosomes), then memory issues may arise. However, this should not be a problem with most datasets. Technically by default QoRTs can run on arbitrarily-ordered bam files, but this is STRONGLY not recommended, as memory usage will by greatly increased. (flag)</p>
</blockquote>

<h3>--verbose:</h3>

<blockquote>
  <p>Flag to indicate that debugging information and extra progress information should be sent to stderr. (flag)</p>
</blockquote>

<h3>--quiet:</h3>

<blockquote>
  <p>Flag to indicate that only errors and warnings should be sent to stderr. (flag)</p>
</blockquote>

<h2>AUTHORS:</h2>

<p>Stephen W. Hartley, Ph.D. stephen.hartley (at nih dot gov)</p>

<h2>LEGAL:</h2>

<p>This software is "United States Government Work" under the terms of the United States Copyright  Act.  It was written as part of the authors' official duties for the United States Government and  thus cannot be copyrighted.  This software is freely available to the public for use without a  copyright notice.  Restrictions cannot be placed on its present or future use.  Although all reasonable efforts have been taken to ensure the accuracy and reliability of the  software and data, the National Human Genome Research Institute (NHGRI) and the U.S. Government  does not and cannot warrant the performance or results that may be obtained by using this software  or data.  NHGRI and the U.S. Government disclaims all warranties as to performance, merchantability  or fitness for any particular purpose.  In any work or product derived from this material, proper attribution of the authors as the source  of the software or data should be made, using "NHGRI Genome Technology Branch" as the citation.  NOTE: This package includes (internally) the sam-1.113.jar library from picard tools. That package uses the MIT license, which can be accessed using the command:  java -jar thisjarfile.jar help samjdkinfo</p>
