<h1>QoRTs: Quality Of Rna-seq Tool Set</h1>

<blockquote>
  <p>Version 0.2.17 (Updated Tue Apr 14 13:54:48 EDT 2015)</p>

<p>(<a href="../index.html">back to main</a>) (<a href="index.html">back to java-utility help</a>)</p>
</blockquote>

<h2>Help for java command "QC"</h2>

<h2>USAGE:</h2>

<pre><code>java [Java Options] -jar QoRTs.jar QC [options] infile gtffile.gtf qcDataDir
</code></pre>

<h2>DESCRIPTION:</h2>

<p>This utility runs a large battery of QC / data processing tools on a single given sam or bam file. This is the primary function of the QoRTs utility. All analyses are run via a single pass through the sam/bam file.</p>

<h2>REQUIRED ARGUMENTS:</h2>

<h3>infile:</h3>

<blockquote>
  <p>The input .bam or .sam file of aligned sequencing reads. Or '-' to read from stdin. (String)</p>
</blockquote>

<h3>gtffile.gtf:</h3>

<blockquote>
  <p>The gtf annotation file. This tool was designed to use the standard gtf annotations provided by Ensembl, but other annotations can be used as well.
If the filename ends with ".gz" or ".zip", the file will be parsed using the appropriate decompression method. (String)</p>
</blockquote>

<h3>qcDataDir:</h3>

<blockquote>
  <p>The output directory. (String)</p>
</blockquote>

<h2>OPTIONAL ARGUMENTS:</h2>

<h3>--singleEnded:</h3>

<blockquote>
  <p>Flag to indicate that reads are single end. (flag)</p>
</blockquote>

<h3>--stranded:</h3>

<blockquote>
  <p>Flag to indicate that data is stranded. (flag)</p>
</blockquote>

<h3>--stranded_fr_secondstrand:</h3>

<blockquote>
  <p>Flag to indicate that reads are from a fr_secondstrand type of stranded library (equivalent to the "stranded = yes" option in HTSeq or the "fr_secondStrand" library-type option in TopHat/CuffLinks). If your data is stranded, you must know the library type in order to analyze it properly. This utility uses the same definitions as cufflinks to define strandedness type. By default, the fr_firststrand library type is assumed for all stranded data (equivalent to the "stranded = D" option in HTSeq). (flag)</p>
</blockquote>

<h3>--maxReadLength len:</h3>

<blockquote>
  <p>Sets the maximum read length. For unclipped datasets this option is not necessary since the read length can be determined from the data. By default, QoRTs will attempt to determine the max read length by examining the first 1000 reads. If your data is hard-clipped prior to alignment, then it is strongly recommended that this option be included, or else an error may occur. Note that hard-clipping data prior to alignment is generally not recommended, because this makes it difficult (or impossible) to determine the sequencer read-cycle of each nucleotide base. This may obfuscate cycle-specific artifacts, trends, or errors, the detection of which is one of the primary purposes of QoRTs! In addition, hard clipping (whether before or after alignment) removes quality score data, and thus quality score metrics may be misleadingly optimistic. A MUCH preferable method of removing undesired sequence is to replace such sequence with N's, which preserves the quality score and the sequencer cycle information while still removing undesired sequence.  (Int)</p>
</blockquote>

<h3>--generatePlots:</h3>

<blockquote>
  <p>Generate all single-replicate QC plots. Equivalent to the combination of: --generateMultiPlot --generateSeparatePlots and --generatePdfReport. This option will cause QoRTs to make an Rscript system call, loading the R package QoRTs. (Note: this requires that R be installed and in the PATH, and that QoRTs be installed on that R installation) (flag)</p>
</blockquote>

<h3>--testRun:</h3>

<blockquote>
  <p>Flag to indicate that only the first 100k reads should be read in. Used for testing. (flag)</p>
</blockquote>

<h3>--minMAPQ num:</h3>

<blockquote>
  <p>Filter out reads with less than the given MAPQ. Set to 0 to turn off mapq filtering. (Int)</p>
</blockquote>

<h3>--keepMultiMapped:</h3>

<blockquote>
  <p>Flag to indicate that the tool should NOT filter out multi-mapped reads. Note that even with this flag raised this utility will still only use the 'primary' alignment location for each read. By default any reads that are marked as multi-mapped will be ignored entirely. Most aligners use the MAPQ value to mark multi-mapped reads. Any read with MAPQ &lt; 255 is assumed to be non-uniquely mapped (this is the standard used by RNA-STAR and TopHat/TopHat2).  This option is equivalent to "--minMAPQ 0". (flag)</p>
</blockquote>

<h3>--noGzipOutput:</h3>

<blockquote>
  <p>Flag to indicate that output files should NOT be compressed into the gzip format. By default almost all output files are compressed to save space. (flag)</p>
</blockquote>

<h3>--readGroup readGroupName:</h3>

<blockquote>
  <p>If this option is set, all analyses will be restricted to ONLY reads that are tagged with the given readGroupName (using an RG tag). This can be used if multiple read-groups have already been combined into a single bam file, but you want to summarize each read-group separately. (String)</p>
</blockquote>

<h3>--dropChrom dropChromosomes:</h3>

<blockquote>
  <p>A comma-delimited list of chromosomes to ignore and exclude from all analyses. Important: no whitespace! (CommaDelimitedListOfStrings)</p>
</blockquote>

<h3>--skipFunctions func1,func2,...:</h3>

<blockquote>
  <p>A comma-delimited list of functions to skip. Important: No whitespace! The default-on functions are: NVC, GCDistribution, GeneCalcs, QualityScoreDistribution, writeKnownSplices, writeNovelSplices, writeClippedNVC, CigarOpDistribution, InsertSize, chromCounts, writeSpliceExon, writeGenewiseGeneBody, JunctionCalcs, writeGeneCounts, writeDESeq, writeDEXSeq, writeGeneBody, StrandCheck (CommaDelimitedListOfStrings)</p>
</blockquote>

<h3>--addFunctions func1,func2,...:</h3>

<blockquote>
  <p>A list of functions to add. This can be used to add functions that are off by default. Followed by a comma delimited list, with no internal whitespace. The default-off functions are: annotatedSpliceExonCounts, FPKM, cigarMatch, makeJunctionBed, makeWiggles, makeAllBrowserTracks (CommaDelimitedListOfStrings)</p>
</blockquote>

<h3>--runFunctions func1,func2,...:</h3>

<blockquote>
  <p>The complete list of functions to run. Setting this option runs ONLY for the functions explicitly requested here (along with any dependancy functions). The list should be formatted as a comma delimited list, with no internal whitespace. Allowed options are: NVC, annotatedSpliceExonCounts, GCDistribution, GeneCalcs, FPKM, cigarMatch, QualityScoreDistribution, writeKnownSplices, writeNovelSplices, writeClippedNVC, CigarOpDistribution, InsertSize, chromCounts, writeSpliceExon, writeGenewiseGeneBody, JunctionCalcs, writeGeneCounts, makeJunctionBed, writeDESeq, writeDEXSeq, makeWiggles, writeGeneBody, StrandCheck, makeAllBrowserTracks (CommaDelimitedListOfStrings)</p>
</blockquote>

<h3>--seqReadCt val:</h3>

<blockquote>
  <p>(Optional) The total number of reads (or read-pairs, for paired-end data) generated by the sequencer for this sample, prior to alignment. This will be passed on into the QC.summary.txt file and used to calculate mapping rate. (Int)</p>
</blockquote>

<h3>--rawfastq myfastq.fq.gz:</h3>

<blockquote>
  <p>(Optional) The raw fastq, prior to alignment. This is used ONLY to calculate the number of pre-alignment reads (or read-pairs) simply by counting the number of lines and dividing by 4. Alternatively, the number of pre-alignment read-pairs can be included explicitly via the --seqReadCt option, or added in the plotting / cross-comparison step by including the input.read.pair.count column in the replicate decoder.In general, the --seqReadCt option is recommended when available.
If the filename ends with ".gz" or ".zip", the file will be parsed using the appropriate decompression method. (String)</p>
</blockquote>

<h3>--chromSizes chrom.sizes.txt:</h3>

<blockquote>
  <p>A chrom.sizes file. The first (tab-delimited) column must contain all chromosomes found in the dataset. The second column must contain chromosome sizes (in base-pairs). If a standard genome is being used, it is strongly recommended that this be generated by the UCSC utility 'fetchChromSizes'.
This file is ONLY needed to produce wiggle files. If this is provided, then by default QoRTs will produce 100-bp-window wiggle files (and junction '.bed' files) for the supplied data.In order to produce wiggle files, this parameter is REQUIRED. (String)</p>
</blockquote>

<h3>--title myTitle:</h3>

<blockquote>
  <p>The title of the replicate. Used for the track name in the track definition line of any browser tracks ('.wig' or '.bed' files) generated by this utility. Also may be used in the figure text, if figures are being generated.Note that no browser tracks will be created by default, unless the '--chromSizes' option is set. Bed files can also be generated using the option '--addFunction makeJunctionBed' (String)</p>
</blockquote>

<h3>--flatgff flattenedGffFile.gff.gz:</h3>

<blockquote>
  <p>A "flattened" gff file that matches the standard gtf file. Optional. The "flattened" gff file assigns unique identifiers for all exons, splice junctions, and aggregate-genes. This is used for the junction counts and exon counts (for DEXSeq). The flattened gtf file can be generated using the "makeFlatGff" command. Flattened GFF files containing novel splice junctions can be generated using the "mergeNovelSplices" function. Note that (for most purposes) the command should be run with the same strandedness code as found in the dataset. Running a flattened gff that was generated using a different strandedness mode may be useful for certain purposes, but is generally not supported and is for advanced users only.See the documentation for makeFlatGff for more information. 
If the filename ends with ".gz" or ".zip", the file will be parsed using the appropriate decompression method. (String)</p>
</blockquote>

<h3>--generateMultiPlot:</h3>

<blockquote>
  <p>Generate a multi-frame figure, containing a visual summary of all QC stats. (Note: this requires that R be installed and in the PATH, and that QoRTs be installed on that R installation) (flag)</p>
</blockquote>

<h3>--generateSeparatePlots:</h3>

<blockquote>
  <p>Generate seperate plots for each QC stat, rather than only one big multiplot. (Note: this requires that R be installed and in the PATH, and that QoRTs be installed on that R installation)  (flag)</p>
</blockquote>

<h3>--generatePdfReport:</h3>

<blockquote>
  <p>Generate a pdf report. (Note: this requires that R be installed and in the PATH, and that QoRTs be installed on that R installation) (flag)</p>
</blockquote>

<h3>--restrictToGeneList geneList.txt:</h3>

<blockquote>
  <p>If this option is set, almost all analyses will be restricted to reads that are found on genes named in the supplied gene list file. The file should contain a gene ID on each line and nothing else. The only functions that will be run on the full set of all reads will be the functions that calculate the gene mapping itself. NOTE: if you want to include ambiguous reads, include a line with the text: '_ambiguous'. If you want to include reads that do not map to any known feature, include a line with the text: '_no_feature'. WARNING: this is not intended for default use. It is intended to be used when re-running QoRTs, with the intention of examining artifacts that can be caused in various plots by a small number of genes with extremely high coverage. For example, GC content plots sometimes contain visible spikes caused by small mitochondrial genes with extremely high expression.ADDITIONAL WARNING: This feature is in BETA, and is not yet fully tested. (String)</p>
</blockquote>

<h3>--dropGeneList geneList.txt:</h3>

<blockquote>
  <p>If this option is set, almost all analyses will be restricted to reads that are NOT found on genes named in the supplied gene list file. The file should contain a gene ID on each line and nothing else. The only functions that will be run on the full set of all reads will be the functions that calculate the gene mapping itself. NOTE: if you want to EXCLUDE ambiguous reads, include a line with the text: '_ambiguous'. If you want to EXCLUDE reads that do not map to any known feature, include a line with the text: '_no_feature'. WARNING: this is not intended for default use. It is intended to be used when re-running QoRTs, with the intention of examining artifacts that can be caused by certain individual 'problem genes'. For example, GC content plots sometimes contain visible spikes caused by small transcripts / RNA's with extremely high expression levels.ADDITIONAL WARNING: This feature is in BETA, and is not yet fully tested. (String)</p>
</blockquote>

<h3>--nameSorted:</h3>

<blockquote>
  <p>Relevant for paired-end reads only. 
This flag is used to run QoRTs in "name-sorted" mode. This flag is optional, as under the default mode QoRTs will accept BAM files sorted by either name OR position. However, The only actual requirement in this mode is that read pairs be adjacent. 
Errors may occur if the SAM flags are inconsistent: for example, if orphaned reads appear with the "mate mapped" SAM flag set. (flag)</p>
</blockquote>

<h3>--coordSorted:</h3>

<blockquote>
  <p>DEPRECIATED: this mode is now subsumed by the default mode and as such this parameter is now nonfunctional.
Note that, in the default mode, for paired-end data QoRTs will accept EITHER coordinate-sorted OR name-sorted bam files. In "--nameSorted" mode, QoRTs ONLY accepts name-sorted bam files.
If a large fraction of the read-pairs are mapped to extremely distant loci (or to different chromosomes), then memory issues may arise. However, this should not be a problem with most datasets. Technically by default QoRTs can run on arbitrarily-ordered bam files, but this is STRONGLY not recommended, as memory usage will by greatly increased. (flag)</p>
</blockquote>

<h3>--fileContainsNoMultiMappedReads:</h3>

<blockquote>
  <p>DEPRECIATED. Flag to indicate that the input sam/bam file contains only primary alignments (ie, no multi-mapped reads). This flag is ALWAYS OPTIONAL, but when applicable this utility will run (slightly) faster when using this argument. (DEPRECIATED! The performance improvement was marginal) (flag)</p>
</blockquote>

<h3>--parallelFileRead:</h3>

<blockquote>
  <p>DEPRECIATED: DO NOT USE. Flag to indicate that bam file reading should be run in paralell for increased speed. Note that in this mode you CANNOT read from stdin. Also note that for this to do anything useful, the numThreads option must be set to some number greater than 1. Also note that additional threads above 9 will have no appreciable affect on speed. (flag)</p>
</blockquote>

<h3>--numThreads num:</h3>

<blockquote>
  <p>DEPRECIATED, nonfunctional. (Int)</p>
</blockquote>

<h3>--verbose:</h3>

<blockquote>
  <p>Flag to indicate that debugging information and extra progress information should be sent to stderr. (flag)</p>
</blockquote>

<h3>--quiet:</h3>

<blockquote>
  <p>Flag to indicate that only errors and warnings should be sent to stderr. (flag)</p>
</blockquote>

<h2>AUTHORS:</h2>

<p>Stephen W. Hartley, Ph.D. stephen.hartley (at nih dot gov)</p>

<h2>LEGAL:</h2>

<p>This software is "United States Government Work" under the terms of the United States Copyright  Act.  It was written as part of the authors' official duties for the United States Government and  thus cannot be copyrighted.  This software is freely available to the public for use without a  copyright notice.  Restrictions cannot be placed on its present or future use.  Although all reasonable efforts have been taken to ensure the accuracy and reliability of the  software and data, the National Human Genome Research Institute (NHGRI) and the U.S. Government  does not and cannot warrant the performance or results that may be obtained by using this software  or data.  NHGRI and the U.S. Government disclaims all warranties as to performance, merchantability  or fitness for any particular purpose.  In any work or product derived from this material, proper attribution of the authors as the source  of the software or data should be made, using "NHGRI Genome Technology Branch" as the citation.  NOTE: This package includes (internally) the sam-1.113.jar library from picard tools. That package uses the MIT license, which can be accessed using the command:  java -jar thisjarfile.jar help samjdkinfo</p>
