<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head><title>R: Nucleotide-rate by read cycle plot</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<link rel="stylesheet" type="text/css" href="R.css">
</head><body>

<table width="100%" summary="page for makePlot.raw.NVC {QoRTs}"><tr><td>makePlot.raw.NVC {QoRTs}</td><td align="right">R Documentation</td></tr></table>

<h2>
Nucleotide-rate by read cycle plot
</h2>

<h3>Description</h3>

<p>Plots the nucleotide rate by read cycle
</p>


<h3>Usage</h3>

<pre>
makePlot.raw.NVC(plotter,  r2.buffer = NULL, 
                         points.highlighted = TRUE,
                         label.majority.bases = FALSE, 
                         label.majority.bases.threshold = 0.5, 
                         label.majority.bases.cex = 0.5, 
                         rasterize.plotting.area = FALSE, 
                         raster.height = 1000, 
                         raster.width = 2000,
                         show.base.legend = TRUE, 
                         debugMode, singleEndMode,
                         ...)
                         
makePlot.minus.clipping.NVC(plotter,  r2.buffer = NULL, 
                         points.highlighted = TRUE,
                         label.majority.bases = FALSE, 
                         label.majority.bases.threshold = 0.5, 
                         label.majority.bases.cex = 0.5, 
                         rasterize.plotting.area = FALSE, 
                         raster.height = 1000, 
                         raster.width = 2000,
                         show.base.legend = TRUE, 
                         debugMode, singleEndMode,
                         ...)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>plotter</code></td>
<td>

<p>A <code>QoRT_Plotter</code> reference object. See <code><a href="build.plotter.html">build.plotter</a></code>.
</p>
</td></tr>
<tr valign="top"><td><code>r2.buffer</code></td>
<td>

<p>Buffer space to place between the plotting of read 1
and read 2. By default this will choose a reasonable value.
</p>
</td></tr>
<tr valign="top"><td><code>points.highlighted</code></td>
<td>

<p>A logical value. Determines whether to ever plot points on top of the lines. This
can be useful for identifying outliers. If TRUE, then all highlighted lane-bams
will be overlaid with points using their designated pch symbol. If the plotter does
not highlight any lanebams, then points will be overlaid on ALL lines.
</p>
</td></tr>
<tr valign="top"><td><code>label.majority.bases</code></td>
<td>

<p>A logical value indicating whether to label the genotypes of
read cycles in which the most common base has a frequency 
exceeding label.majority.bases.threshold (see below).
</p>
</td></tr>
<tr valign="top"><td><code>label.majority.bases.threshold</code></td>
<td>

<p>A numeric value indicating the cutoff above which the most 
common base will be labelled, if label.majority.bases is set
TRUE (see above).
</p>
</td></tr>
<tr valign="top"><td><code>label.majority.bases.cex</code></td>
<td>

<p>The cex value fed to text() that is used to determine how
big the labels are to be, if label.majority.bases is TRUE.
(see <code><a href="../../graphics/html/par.html">par</a></code> for information on cex).
</p>
</td></tr>
<tr valign="top"><td><code>rasterize.plotting.area</code></td>
<td>

<p>Logical. If TRUE, the plotting area will be written to a temp
png file then drawn to the current device as a raster image.
This option is generally preferred for vector devices, because
NVC plots can be very large when drawn in vector format. 
<EM>Note: this requires the <code>png</code> package!</EM>
</p>
</td></tr>
<tr valign="top"><td><code>raster.height</code></td>
<td>

<p>Numeric. If rasterize.plotting.area is TRUE, then this is the
height of the plotting area raster image, in pixels.
</p>
</td></tr>
<tr valign="top"><td><code>raster.width</code></td>
<td>

<p>Numeric. If rasterize.plotting.area is TRUE, then this is the
width of the plotting area raster image, in pixels.
</p>
</td></tr>
<tr valign="top"><td><code>show.base.legend</code></td>
<td>

<p>A logical value indicating whether to print the base-color
legend along the right edge of the plot.
</p>
</td></tr>
<tr valign="top"><td><code>debugMode</code></td>
<td>

<p>Logical. If TRUE, debugging data will be printed to the console.
</p>
</td></tr>
<tr valign="top"><td><code>singleEndMode</code></td>
<td>

<p>Logical. Determines whether the dataset consists of single-ended reads. 
By default this is determined from the data. Thus, this parameter should 
never need to be set by the user.
</p>
</td></tr>
<tr valign="top"><td><code>...</code></td>
<td>
 
<p>Arguments to be passed to methods, such as
<a href="../../graphics/html/par.html">graphical parameters</a> (see <code><a href="../../graphics/html/par.html">par</a></code>).
</p>
</td></tr>
</table>


<h3>Details</h3>

<p>This plot displays the nucleotide rates by read cycle. The color scheme for NVC plots is different from all other plots. Rather than being used for emphasis or to allow cross-comparisons by sample, biological-condition, or lane, the colors are used to indicate the four nucleotides: A (green), T (red), G (orange), or C (blue). Depending on the type of QoRT_Plotter being used, sample-runs will be marked and differentiated by marking the lines with shapes (R points). In many cases the points will be unreadable due to overplotting, but clear outliers that stray from the general trends can be readily identified.
</p>
<p>When used with a &quot;sample.highlight&quot; type QoRT_Plotter (see <code><a href="build.plotter.html">build.plotter</a></code>), &quot;highlighted&quot; samples will be drawn with a deeper shade of the given color.
</p>
<p><code><a href="makePlot.raw.NVC.html">makePlot.raw.NVC</a></code> plots the nucleotides of all cycles of all aligned reads. <code><a href="makePlot.raw.NVC.html">makePlot.minus.clipping.NVC</a></code> plots the nucleotides for all cycles that are NOT soft-clipped by the aligner. Note that for reads aligned with an aligner that does not perform soft-clipping (such as TopHat2), the two plots will be identical.
</p>


<h3>See Also</h3>

<p><code><a href="build.plotter.html">build.plotter</a></code>
</p>

<hr><div align="center">[Package <em>QoRTs</em> version 0.1.6 <a href="00Index.html">Index</a>]</div>
</body></html>
