<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head><title>R: Clipped sequence nucleotide-by-position plot</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<link rel="stylesheet" type="text/css" href="R.css">
</head><body>

<table width="100%" summary="page for makePlot.NVC.lead.clip {QoRTs}"><tr><td>makePlot.NVC.lead.clip {QoRTs}</td><td align="right">R Documentation</td></tr></table>

<h2>
Clipped sequence nucleotide-by-position plot
</h2>

<h3>Description</h3>

<p>Plots the nucleotide rates for clipped segments
</p>


<h3>Usage</h3>

<pre>
makePlot.NVC.lead.clip(plotter, clip.amt,  r2.buffer, 
                   points.highlighted = TRUE,
                   label.majority.bases = TRUE, 
                   label.majority.bases.threshold = 0.5, 
                   label.majority.bases.cex = 1, 
                   rasterize.plotting.area = FALSE, 
                   raster.height = 1000, 
                   raster.width = 1000,
                   show.base.legend = TRUE, 
                   debugMode, singleEndMode,
                   ...)


makePlot.NVC.tail.clip(plotter, clip.amt,  r2.buffer, 
                       points.highlighted = TRUE, 
                       label.majority.bases = TRUE, 
                       label.majority.bases.threshold = 0.5, 
                       label.majority.bases.cex = 1, 
                       rasterize.plotting.area = FALSE, 
                       raster.height = 1000, 
                       raster.width = 1000,
                       show.base.legend = TRUE, 
                       debugMode, singleEndMode, ...)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>plotter</code></td>
<td>

<p>A <code>QoRT_Plotter</code> reference object. See <code><a href="build.plotter.html">build.plotter</a></code>.
</p>
</td></tr>
<tr valign="top"><td><code>clip.amt</code></td>
<td>

<p>Numeric value. The number of bases clipped. These methods
only plot the sequence for a single specific clip.amt. 
</p>
</td></tr>
<tr valign="top"><td><code>r2.buffer</code></td>
<td>

<p>Buffer space to place between the plotting of read 1
and read 2. By default this will choose a reasonable value.
</p>
</td></tr>
<tr valign="top"><td><code>points.highlighted</code></td>
<td>

<p>A logical value. Determines whether to ever plot points on top of the lines. This
can be useful for identifying outliers. If TRUE, then all highlighted lane-bams
will be overlaid with points using their designated pch symbol. If the plotter does
not highlight any lanebams, then points will be overlaid on ALL lines.
</p>
</td></tr>
<tr valign="top"><td><code>label.majority.bases</code></td>
<td>

<p>A logical value indicating whether to label the genotypes of
read cycles in which the most common base has a frequency 
exceeding label.majority.bases.threshold (see below).
</p>
</td></tr>
<tr valign="top"><td><code>label.majority.bases.threshold</code></td>
<td>

<p>A numeric value indicating the cutoff above which the most 
common base will be labelled, if label.majority.bases is set
TRUE (see above).
</p>
</td></tr>
<tr valign="top"><td><code>label.majority.bases.cex</code></td>
<td>

<p>The cex value fed to text() that is used to determine how
big the labels are to be, if label.majority.bases is TRUE.
(see <code><a href="../../graphics/html/par.html">par</a></code> for information on cex).
</p>
</td></tr>
<tr valign="top"><td><code>rasterize.plotting.area</code></td>
<td>

<p>Logical. If TRUE, the plotting area will be written to a temp
png file then drawn to the current device as a raster image.
This option is generally preferred for vector devices, because
NVC plots can be very large when drawn in vector format. 
<EM>Note: this requires the <code>png</code> package!</EM>
</p>
</td></tr>
<tr valign="top"><td><code>raster.height</code></td>
<td>

<p>Numeric. If rasterize.plotting.area is TRUE, then this is the
height of the plotting area raster image, in pixels.
</p>
</td></tr>
<tr valign="top"><td><code>raster.width</code></td>
<td>

<p>Numeric. If rasterize.plotting.area is TRUE, then this is the
width of the plotting area raster image, in pixels.
</p>
</td></tr>
<tr valign="top"><td><code>show.base.legend</code></td>
<td>

<p>A logical value indicating whether to print the base-color
legend along the right edge of the plot.
</p>
</td></tr>
<tr valign="top"><td><code>debugMode</code></td>
<td>

<p>Logical. If TRUE, debugging data will be printed to the console.
</p>
</td></tr>
<tr valign="top"><td><code>singleEndMode</code></td>
<td>

<p>Logical. Determines whether the dataset consists of single-ended reads. 
By default this is determined from the data. Thus, this parameter should 
never need to be set by the user.
</p>
</td></tr>
<tr valign="top"><td><code>...</code></td>
<td>
 
<p>Arguments to be passed to methods, such as
<a href="../../graphics/html/par.html">graphical parameters</a> (see <code><a href="../../graphics/html/par.html">par</a></code>).
</p>
</td></tr>
</table>


<h3>Details</h3>

<p>For general information on reading NVC plots, see <code><a href="makePlot.raw.NVC.html">makePlot.NVC</a></code>.
</p>
<p>This function plots the nucleotide rates for the sections of reads that were soft-clipped by the aligner. Note that these will not function on reads that have been aligned using an aligner that does not practice soft clipping (such as TopHat2).
</p>


<h3>See Also</h3>

<p><code><a href="build.plotter.html">build.plotter</a></code>, <code><a href="makePlot.raw.NVC.html">makePlot.NVC</a></code>
</p>

<hr><div align="center">[Package <em>QoRTs</em> version 1.0.1 <a href="00Index.html">Index</a>]</div>
</body></html>
