\batchmode


\documentclass[12pt]{article}
\RequirePackage{ifthen}

\usepackage{graphicx}\usepackage[usenames,dvipsnames]{color}
\makeatletter

\makeatother


\definecolor{fgcolor}{rgb}{0.345, 0.345, 0.345}%
\providecommand{\hlnum}[1]{\textcolor[rgb]{0.686,0.059,0.569}{#1}}%%
\providecommand{\hlstr}[1]{\textcolor[rgb]{0.192,0.494,0.8}{#1}}%%
\providecommand{\hlcom}[1]{\textcolor[rgb]{0.678,0.584,0.686}{\textit{#1}}}%%
\providecommand{\hlopt}[1]{\textcolor[rgb]{0,0,0}{#1}}%%
\providecommand{\hlstd}[1]{\textcolor[rgb]{0.345,0.345,0.345}{#1}}%%
\providecommand{\hlkwa}[1]{\textcolor[rgb]{0.161,0.373,0.58}{\textbf{#1}}}%%
\providecommand{\hlkwb}[1]{\textcolor[rgb]{0.69,0.353,0.396}{#1}}%%
\providecommand{\hlkwc}[1]{\textcolor[rgb]{0.333,0.667,0.333}{#1}}%%
\providecommand{\hlkwd}[1]{\textcolor[rgb]{0.737,0.353,0.396}{\textbf{#1}}}%
 \let\hlipl\textcolor[rgb]{0.69,0.353,0.396}{
 }
\usepackage{framed}
\makeatletter

%
\newenvironment{kframe}{%
 \def \at@end@of@kframe{}%
 \ifinner\ifhmode \def \at@end@of@kframe{\end{minipage}}%
  \begin{minipage}{\columnwidth}%
 \fi\fi \def \FrameCommand##1{\hskip\@totalleftmargin \hskip-\fboxsep
 \colorbox{shadecolor}{##1}\hskip-\fboxsep
     % There is no \\@totalrightmargin, so:
     \hskip-\linewidth \hskip-\@totalleftmargin \hskip\columnwidth}%
 \MakeFramed {\advance\hsize-\width
   \@totalleftmargin\z@ \linewidth\hsize
   \@setminipage}}%
 {

\unskip\endMakeFramed \at@end@of@kframe} 
\makeatother


\definecolor{shadecolor}{rgb}{.97, .97, .97}
\definecolor{messagecolor}{rgb}{0, 0, 0}
\definecolor{warningcolor}{rgb}{1, 0, 1}
\definecolor{errorcolor}{rgb}{1, 0, 0}

%
\newenvironment{knitrout}{}{}  % an empty environment to be redefined in TeX


\usepackage{alltt}


\RequirePackage{/mnt/nfs/gigantor/ifs/DCEG/Home/hartleys/R/x86_64-pc-linux-gnu-library/3.3/BiocStyle/resources/tex/Bioconductor}


\AtBeginDocument{\bibliographystyle{/mnt/nfs/gigantor/ifs/DCEG/Home/hartleys/R/x86_64-pc-linux-gnu-library/3.3/BiocStyle/resources/tex/unsrturl}}


\usepackage[sort]{cite}
\usepackage{hyperref}
\hypersetup{bookmarksopen=false}

%
\providecommand{\myincfigRaw}[3]{%
  \begin{figure}[h]
  \centering
  \includegraphics[width=#2]{#1}
  \caption{\label{figure/#1}#3}
   \label{#1-fig}
  \end{figure}
}%
\providecommand{\myincfig}[3]{%
  \begin{figure}[h]
  \centering
  \includegraphics[width=#2]{figure/#1-1}
  \caption{\label{figure/#1}#3}
  \label{#1-fig}
  \end{figure}
}%
\providecommand{\myincfigTwo}[3]{%
  \begin{figure}[h]
  \centering
  \includegraphics[width=#2]{figure/#1-1}
  \includegraphics[width=#2]{figure/#1-2}  
  \caption{\label{figure/#1}#3}
  \label{#1-fig}
  \end{figure}
}%
\providecommand{\myincfigThree}[3]{%
  \begin{figure}[h]
  \centering
  \includegraphics[width=#2]{figure/#1-1}
  \includegraphics[width=#2]{figure/#1-2}  
  \includegraphics[width=#2]{figure/#1-3}  
  \caption{\label{figure/#1}#3}
  \label{#1-fig}
  \end{figure}
} 


\let\olditemize
  \olditemize
  
\setlength{\itemsep}{1pt} 
  
\setlength{\parskip}{0pt} 
  
\setlength{\parsep}{0pt} 

%
\renewcommand{\itemize}{
  \olditemize
  \setlength{\itemsep}{1pt}
  \setlength{\parskip}{0pt}
  \setlength{\parsep}{0pt}
}


\title{QoRTs Package User Manual}
\author{Stephen Hartley\\
National Human Genome Research Institute\\
National Institutes of Health
}
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}



\pagecolor[gray]{.7}

\usepackage[]{inputenc}



\makeatletter

\makeatletter
\count@=\the\catcode`\_ \catcode`\_=8 
\newenvironment{tex2html_wrap}{}{}%
\catcode`\<=12\catcode`\_=\count@
\newcommand{\providedcommand}[1]{\expandafter\providecommand\csname #1\endcsname}%
\newcommand{\renewedcommand}[1]{\expandafter\providecommand\csname #1\endcsname{}%
  \expandafter\renewcommand\csname #1\endcsname}%
\newcommand{\newedenvironment}[1]{\newenvironment{#1}{}{}\renewenvironment{#1}}%
\let\newedcommand\renewedcommand
\let\renewedenvironment\newedenvironment
\makeatother
\let\mathon=$
\let\mathoff=$
\ifx\AtBeginDocument\undefined \newcommand{\AtBeginDocument}[1]{}\fi
\newbox\sizebox
\setlength{\hoffset}{0pt}\setlength{\voffset}{0pt}
\addtolength{\textheight}{\footskip}\setlength{\footskip}{0pt}
\addtolength{\textheight}{\topmargin}\setlength{\topmargin}{0pt}
\addtolength{\textheight}{\headheight}\setlength{\headheight}{0pt}
\addtolength{\textheight}{\headsep}\setlength{\headsep}{0pt}
\setlength{\textwidth}{349pt}
\newwrite\lthtmlwrite
\makeatletter
\let\realnormalsize=\normalsize
\global\topskip=2sp
\def\preveqno{}\let\real@float=\@float \let\realend@float=\end@float
\def\@float{\let\@savefreelist\@freelist\real@float}
\def\liih@math{\ifmmode$\else\bad@math\fi}
\def\end@float{\realend@float\global\let\@freelist\@savefreelist}
\let\real@dbflt=\@dbflt \let\end@dblfloat=\end@float
\let\@largefloatcheck=\relax
\let\if@boxedmulticols=\iftrue
\def\@dbflt{\let\@savefreelist\@freelist\real@dbflt}
\def\adjustnormalsize{\def\normalsize{\mathsurround=0pt \realnormalsize
 \parindent=0pt\abovedisplayskip=0pt\belowdisplayskip=0pt}%
 \def\phantompar{\csname par\endcsname}\normalsize}%
\def\lthtmltypeout#1{{\let\protect\string \immediate\write\lthtmlwrite{#1}}}%
\newcommand\lthtmlhboxmathA{\adjustnormalsize\setbox\sizebox=\hbox\bgroup\kern.05em }%
\newcommand\lthtmlhboxmathB{\adjustnormalsize\setbox\sizebox=\hbox to\hsize\bgroup\hfill }%
\newcommand\lthtmlvboxmathA{\adjustnormalsize\setbox\sizebox=\vbox\bgroup %
 \let\ifinner=\iffalse \let\)\liih@math }%
\newcommand\lthtmlboxmathZ{\@next\next\@currlist{}{\def\next{\voidb@x}}%
 \expandafter\box\next\egroup}%
\newcommand\lthtmlmathtype[1]{\gdef\lthtmlmathenv{#1}}%
\newcommand\lthtmllogmath{\dimen0\ht\sizebox \advance\dimen0\dp\sizebox
  \ifdim\dimen0>.95\vsize
   \lthtmltypeout{%
*** image for \lthtmlmathenv\space is too tall at \the\dimen0, reducing to .95 vsize ***}%
   \ht\sizebox.95\vsize \dp\sizebox\z@ \fi
  \lthtmltypeout{l2hSize %
:\lthtmlmathenv:\the\ht\sizebox::\the\dp\sizebox::\the\wd\sizebox.\preveqno}}%
\newcommand\lthtmlfigureA[1]{\let\@savefreelist\@freelist
       \lthtmlmathtype{#1}\lthtmlvboxmathA}%
\newcommand\lthtmlpictureA{\bgroup\catcode`\_=8 \lthtmlpictureB}%
\newcommand\lthtmlpictureB[1]{\lthtmlmathtype{#1}\egroup
       \let\@savefreelist\@freelist \lthtmlhboxmathB}%
\newcommand\lthtmlpictureZ[1]{\hfill\lthtmlfigureZ}%
\newcommand\lthtmlfigureZ{\lthtmlboxmathZ\lthtmllogmath\copy\sizebox
       \global\let\@freelist\@savefreelist}%
\newcommand\lthtmldisplayA{\bgroup\catcode`\_=8 \lthtmldisplayAi}%
\newcommand\lthtmldisplayAi[1]{\lthtmlmathtype{#1}\egroup\lthtmlvboxmathA}%
\newcommand\lthtmldisplayB[1]{\edef\preveqno{(\theequation)}%
  \lthtmldisplayA{#1}\let\@eqnnum\relax}%
\newcommand\lthtmldisplayZ{\lthtmlboxmathZ\lthtmllogmath\lthtmlsetmath}%
\newcommand\lthtmlinlinemathA{\bgroup\catcode`\_=8 \lthtmlinlinemathB}
\newcommand\lthtmlinlinemathB[1]{\lthtmlmathtype{#1}\egroup\lthtmlhboxmathA
  \vrule height1.5ex width0pt }%
\newcommand\lthtmlinlineA{\bgroup\catcode`\_=8 \lthtmlinlineB}%
\newcommand\lthtmlinlineB[1]{\lthtmlmathtype{#1}\egroup\lthtmlhboxmathA}%
\newcommand\lthtmlinlineZ{\egroup\expandafter\ifdim\dp\sizebox>0pt %
  \expandafter\centerinlinemath\fi\lthtmllogmath\lthtmlsetinline}
\newcommand\lthtmlinlinemathZ{\egroup\expandafter\ifdim\dp\sizebox>0pt %
  \expandafter\centerinlinemath\fi\lthtmllogmath\lthtmlsetmath}
\newcommand\lthtmlindisplaymathZ{\egroup %
  \centerinlinemath\lthtmllogmath\lthtmlsetmath}
\def\lthtmlsetinline{\hbox{\vrule width.1em \vtop{\vbox{%
  \kern.1em\copy\sizebox}\ifdim\dp\sizebox>0pt\kern.1em\else\kern.3pt\fi
  \ifdim\hsize>\wd\sizebox \hrule depth1pt\fi}}}
\def\lthtmlsetmath{\hbox{\vrule width.1em\kern-.05em\vtop{\vbox{%
  \kern.1em\kern0.8 pt\hbox{\hglue.17em\copy\sizebox\hglue0.8 pt}}\kern.3pt%
  \ifdim\dp\sizebox>0pt\kern.1em\fi \kern0.8 pt%
  \ifdim\hsize>\wd\sizebox \hrule depth1pt\fi}}}
\def\centerinlinemath{%
  \dimen1=\ifdim\ht\sizebox<\dp\sizebox \dp\sizebox\else\ht\sizebox\fi
  \advance\dimen1by.5pt \vrule width0pt height\dimen1 depth\dimen1 
 \dp\sizebox=\dimen1\ht\sizebox=\dimen1\relax}

\def\lthtmlcheckvsize{\ifdim\ht\sizebox<\vsize 
  \ifdim\wd\sizebox<\hsize\expandafter\hfill\fi \expandafter\vfill
  \else\expandafter\vss\fi}%
\providecommand{\selectlanguage}[1]{}%
\makeatletter \tracingstats = 1 


\begin{document}
\pagestyle{empty}\thispagestyle{empty}\lthtmltypeout{}%
\lthtmltypeout{latex2htmlLength hsize=\the\hsize}\lthtmltypeout{}%
\lthtmltypeout{latex2htmlLength vsize=\the\vsize}\lthtmltypeout{}%
\lthtmltypeout{latex2htmlLength hoffset=\the\hoffset}\lthtmltypeout{}%
\lthtmltypeout{latex2htmlLength voffset=\the\voffset}\lthtmltypeout{}%
\lthtmltypeout{latex2htmlLength topmargin=\the\topmargin}\lthtmltypeout{}%
\lthtmltypeout{latex2htmlLength topskip=\the\topskip}\lthtmltypeout{}%
\lthtmltypeout{latex2htmlLength headheight=\the\headheight}\lthtmltypeout{}%
\lthtmltypeout{latex2htmlLength headsep=\the\headsep}\lthtmltypeout{}%
\lthtmltypeout{latex2htmlLength parskip=\the\parskip}\lthtmltypeout{}%
\lthtmltypeout{latex2htmlLength oddsidemargin=\the\oddsidemargin}\lthtmltypeout{}%
\makeatletter
\if@twoside\lthtmltypeout{latex2htmlLength evensidemargin=\the\evensidemargin}%
\else\lthtmltypeout{latex2htmlLength evensidemargin=\the\oddsidemargin}\fi%
\lthtmltypeout{}%
\makeatother
\setcounter{page}{1}
\onecolumn

% !!! IMAGES START HERE !!!

\stepcounter{section}
\stepcounter{section}
\stepcounter{section}
\stepcounter{subsection}
\stepcounter{section}
\stepcounter{subsection}
\stepcounter{subsection}
\stepcounter{section}
\stepcounter{section}
\stepcounter{subsection}
\stepcounter{section}
\stepcounter{subsection}
\stepcounter{section}
\stepcounter{subsection}
\stepcounter{subsection}
\stepcounter{subsection}
\stepcounter{subsubsection}
\stepcounter{subsubsection}
\stepcounter{subsubsection}
\stepcounter{subsubsection}
\stepcounter{subsubsection}
\stepcounter{subsubsection}
\stepcounter{subsection}
{\newpage\clearpage
\lthtmlfigureA{framed838}%
\begin{framed}
\emph{What it means and what to look for:} 
\par
In general, when examining these plots, users should scan for a number of potential anomalies:
  \begin{itemize} \small
    \item \textbf{"Spikes"}: In which one of the metrics jumps up or down abruptly, then returns to baseline.
    \item \textbf{"Shelves"}: In which one of the metrics jumps up or down abruptly, then continues at an increased or decreased level.
    \item \textbf{Outliers}: In which one or more particular samples or replicates are very different from all or most of the others. The presence of outliers may be an indicator for sample-collection, library-prep, or sequencing errors or artifacts.
    \item \textbf{Systematic Biases}: In which consistent differences appear between subsets of the data (eg. "lane.ID" or "group.ID"). Many of the biases measured by QoRTs are well-characterized, and many downstream analysis tools are robust against them when they are consistent and uniform. However: biases that vary disproportionately between sample groups may still drive false associations downstream.
    \item \textbf{Inconsistent samples}: In which the technical replicates of a specific biological sample shows substantial variation. In most studies, technical variation is very small relative to biological variation. In the example dataset (for example) the technical replicates are plotted almost on top of one another across many of the plots. If technical replicates do not cluster tightly, or if they cluster with the wrong replicates, then this may be an indicator of a sample swap.
  \end{itemize}
\par
Some "anomalous" metrics may be fundamental to the dataset and may not be 
indicative of any quality issues. For example, when profiling two different 
cell types one would expect the two groups to have very different profiles 
across a number of metrics. However: a single sample that is wildly different 
from the others within the same group may be cause for concern. In many cases 
variations may be observed across multiple metrics, all driven by the same 
underlying phenomenon. The breadth and depth of the metrics provided are 
intended to provide the tools necessary to identify the most likely underlying 
source(s) of the aberration(s).
\par
Some aberrations may not be relevant to the study analysis even when they are
representative of a real data quality issue. A moderate increase in the deletion rate
may not have a noticable impact on expression quantification in a simple differential expression
study. The same issue, however, might be catastrophic in a study focused on quantifying the rate of 
RNA transcription errors or RNA editing events. The number of combinations of study design, sample set structure, sequencing technology, and observed 
quality control issues are myriad, and all of these factors would inform decisionmaking. 
\par
Ultimatiely, bioinformaticians must use their own judgement in deciding how to proceed when 
unexplained abnormalities are discovered in their dataset.
\end{framed}%
\lthtmlfigureZ
\lthtmlcheckvsize\clearpage}

\stepcounter{subsubsection}
{\newpage\clearpage
\lthtmlfigureA{framed886}%
\begin{framed} \small
\emph{What it means and what to look for:} These plots can be used to detect sequencer problems, bad lanes, or similar hardware-level artifacts and errors. Look for spikes or shelves, and ensure that the quality score is relatively consistent across samples and lanes, and that any differences that do exist are not disproportionate with respect to the study condition (or "group.ID").
\end{framed}%
\lthtmlfigureZ
\lthtmlcheckvsize\clearpage}

\stepcounter{subsubsection}
{\newpage\clearpage
\lthtmlfigureA{framed903}%
\begin{framed} \small
\emph{What it means and what to look for:} GC bias has been indicated as a potential driver of false discoveries. Under certain circumstances the GC bias may vary by batch or by sample. If this is apparent in your dataset, particularly if it is associated with study conditions, one may need to apply a GC bias correction method (such as CQN).
\end{framed}%
\lthtmlfigureZ
\lthtmlcheckvsize\clearpage}

\stepcounter{subsubsection}
{\newpage\clearpage
\lthtmlfigureA{framed921}%
\begin{framed} \small
\emph{What it means and what to look for:} Abnormalities ("spikes", "shelves", or cases where sample groups are visibly different) in these plots can be caused by adaptor sequencing, gene fusions, mutations, population stratification, or differences in insert size.
\end{framed}%
\lthtmlfigureZ
\lthtmlcheckvsize\clearpage}

\stepcounter{subsubsection}
{\newpage\clearpage
\lthtmlfigureA{framed948}%
\begin{framed}
% latex2html id marker 948
 \small
\emph{What it means and what to look for:} These plots are most 
often used in conjuction with the plots in Section \ref{subsec:plotDesc-8}. 
Among other things these plots can reveal sequencer errors: a sequencer cycle-skip might 
result in a spike in the deletion rate at a particular cycle, whereas an 
incomplete wash may result in a spike in the insertion rate. These plots 
may also reveal biological differences like population stratification 
(eg: a sub-population disproportionately mismatches the reference genome), 
or broad genome rearrangements/editing (eg: in cancer cells).
\par
In the example dataset, apparent spikes are plainly visible, stemming from
deletions in one short and highly-expressed mitochondrial gene. The plots are also fairly
noisy due to the small number of reads used in the example dataset and the extremely low
frequency of deletions/insertions.
\end{framed}%
\lthtmlfigureZ
\lthtmlcheckvsize\clearpage}

\stepcounter{subsubsection}
{\newpage\clearpage
\lthtmlfigureA{framed972}%
\begin{framed}
% latex2html id marker 972
 \small
\emph{What it means and what to look for:} These plots are most often used in 
conjuction with the plots in Section \ref{subsec:plotDesc-4}. They can elucidate 
the nature of any oddities observed in the previous plots. For example: a large 
spike at one particular length may suggest that an apparent spike may be due 
simply to an unannotated variant in one particular high-expression gene, although 
further investigation is likely merited, to confirm that this is the case.
\end{framed}%
\lthtmlfigureZ
\lthtmlcheckvsize\clearpage}

\stepcounter{subsubsection}
{\newpage\clearpage
\lthtmlfigureA{framed995}%
\begin{framed} \small
\emph{What it means and what to look for:} Spikes in the insert size are common, 
generally the result of short, highly-expressed (often mitochondrial) transcripts. 
Because the size-selection of many RNA-Seq protocols is somewhat random, it is 
important to ensure that the resultant size selection is (relatively) consistent, 
and that variations are not associated with study condition status. If one study 
group has disproportionately high or low insert size, this could cause fragment 
bias that could drive the false discovery of differential effects.
\end{framed}%
\lthtmlfigureZ
\lthtmlcheckvsize\clearpage}

\stepcounter{subsubsection}
{\newpage\clearpage
\lthtmlfigureA{framed1012}%
\begin{framed} \small
\emph{What it means and what to look for:} A number of potential sequencer issues 
can cause an abrupt "spike" or "shelf" in this plot. In one real sample assessed 
by QoRTs by the software author, it was determined that the sequencer camera was 
slightly offset on one specific cycle of one specific run. All the reads at the 
bottom or right edges were lost from that cycle forward, causing the rate of "N" 
calls to increase more than a hundred-fold. Once the problem was recognized the 
affected reads were identified and removed.
\end{framed}%
\lthtmlfigureZ
\lthtmlcheckvsize\clearpage}

\stepcounter{subsubsection}
{\newpage\clearpage
\lthtmlfigureA{framed1042}%
\begin{framed} \small
\emph{What it means and what to look for:} When run on degraded RNA and/or when using poly-A selection, RNA-Seq often tends to have "3' bias", in which read coverage is higher on the 3' end of transcripts. The degree of this 3' bias tends to be dependent on the degree of degradation. Many analysis tools are robust against this issue when it occurs uniformly across the dataset; However, if some samples are substantially more degraded than others then this may cause problems downstream, particularly if RNA degradation is associated to the experimental condition(s). When studying this plot, check to make sure the gene-body coverage is consistant and/or matches your expectations.
\par
Note that the overall gene-body coverage may be strongly influenced by extreme high-coverage genes (and potentially sequence-specific biases on those specific transcripts). Therefore the upper-middle-quartile plot is generally the preferred general metric for assessing overall gene-body coverage.
\end{framed}%
\lthtmlfigureZ
\lthtmlcheckvsize\clearpage}

\stepcounter{subsubsection}
{\newpage\clearpage
\lthtmlfigureA{framed1059}%
\begin{framed} \small
\emph{What it means and what to look for:} This plot can reveal a number of phenomena. First of all: if the top few genes dominate a sample (representing a large percentage of the total reads), oddities may appear in many of the other plots produced by QoRTs, as traits specific to these particular genes are dominant over the variation found across the genome. 
\par
It can also reveal biological variations: Different cell types or cells that are healthy, dying, or under stress often have very different diversity profiles from one another. If one sample of within a group is an extreme outlier it may suggest that something is wrong with that sample. 
\par
Or technical issues: These plots will often reveal inefficiency of hemoglobin or ribosome depletion protocols, and can also clearly reveal low library complexity (indicated by a very small number of genes being represented).
\end{framed}%
\lthtmlfigureZ
\lthtmlcheckvsize\clearpage}

\stepcounter{subsubsection}
{\newpage\clearpage
\lthtmlfigureA{framed1077}%
\begin{framed} \small
\emph{What it means and what to look for:} This can reveal sequence-specific biases such as hexamer or primer bias. Additionally, it can reveal adaptor sequencing. Such issues are generally not a problem as long as they are consistent across samples and groups.
\end{framed}%
\lthtmlfigureZ
\lthtmlcheckvsize\clearpage}

\stepcounter{subsubsection}
{\newpage\clearpage
\lthtmlfigureA{framed1097}%
\begin{framed} \small
\emph{What it means and what to look for:} This can reveal sequence-specific biases such as hexamer or primer bias. Such issues are generally not a problem as long as they are consistent across samples and groups. Unlike the raw NVC plot, adaptor sequence will generally be absent from this plot as it usually will not align to the reference.
\end{framed}%
\lthtmlfigureZ
\lthtmlcheckvsize\clearpage}

\stepcounter{subsubsection}
{\newpage\clearpage
\lthtmlfigureA{framed1126}%
\begin{framed} \small
\emph{What it means and what to look for:} If a large proportion of the reads are shorter than the read length then this can reveal the adaptor sequence.
\end{framed}%
\lthtmlfigureZ
\lthtmlcheckvsize\clearpage}

\stepcounter{subsubsection}
\stepcounter{subsubsection}
{\newpage\clearpage
\lthtmlfigureA{framed1186}%
\begin{framed} \small
\emph{What it means and what to look for:} Outliers in these plots can indicate biological variations or the presence of large mapping problems. They may also suggest the presence of large, highly-expressed, unannotated transcripts or genes.
\end{framed}%
\lthtmlfigureZ
\lthtmlcheckvsize\clearpage}

\stepcounter{subsubsection}
{\newpage\clearpage
\lthtmlfigureA{framed1207}%
\begin{framed} \small
\emph{What it means and what to look for:} This plot can be used to detect a number of anomalies. For example: whether mapping or sequencing artifacts caused a disproportionate discovery of novel splice junctions in one sample or batch. It can also be used as an indicator of the comprehensiveness the genome annotation. Replicates that are obvious outliers may have sequencing/technical issues causing false detection of splice junctions.
\par
Abnormalities in the splice junction rates are generally a symptom of larger issues which will generally be picked up by other metrics. Numerous factors can reduce the efficacy by which aligners map across splice junctions, and as such these plots become very important if the intended downstream analyses include transcript assembly, transcript deconvolution, differential splicing, or any other form of analysis that in some way involves the splice junctions themselves. These plots can be used to assess whether other minor abnormalities observed in the other plots are of sufficient severity to impact splice junction mapping and thus potentially compromise such analyses.
\end{framed}%
\lthtmlfigureZ
\lthtmlcheckvsize\clearpage}

\stepcounter{subsubsection}
{\newpage\clearpage
\lthtmlfigureA{framed1225}%
\begin{framed}
% latex2html id marker 1225
 \small
\emph{What it means and what to look for:} This plot is useful for identifying mapping and/or annotation issues, and can indicate the comprehensiveness the genome annotation. Replicates that are obvious outliers may have sequencing/technical issues causing false detection of splice junctions. 
\par
In general, abnormalities in the splice junction rates are generally a symptom of larger issues which will often be picked up by other metrics. See Section \ref{subsec:plotDesc-13}.
\end{framed}%
\lthtmlfigureZ
\lthtmlcheckvsize\clearpage}

\stepcounter{subsubsection}
{\newpage\clearpage
\lthtmlfigureA{framed1244}%
\begin{framed}
% latex2html id marker 1244
 \small
\emph{What it means and what to look for:} This plot is used to detect whether sample-specific or batch effects have a substantial or biased effect on splice junction appearance, either due to differences in the original RNA, or due to artifacts that alter the rate at which the aligner maps across splice junctions. It can assist in identifying mapping and/or annotation issues, and can indicate the comprehensiveness the genome annotation. Replicates that are obvious outliers may have sequencing/technical issues causing false detection of splice junctions.  
\par
In general, abnormalities in the splice junction rates are generally a symptom of larger issues which will often be picked up by other metrics. See Section \ref{subsec:plotDesc-13}.
\end{framed}%
\lthtmlfigureZ
\lthtmlcheckvsize\clearpage}

\stepcounter{subsubsection}
{\newpage\clearpage
\lthtmlfigureA{framed1262}%
\begin{framed}
% latex2html id marker 1262
 \small
\emph{What it means and what to look for:} This plot is used to detect whether sample-specific or batch effects have a substantial or biased effect on splice junction appearance, either due to differences in the original RNA, or due to artifacts that alter the rate at which the aligner maps across splice junctions. This plot is useful for identifying mapping and/or annotation issues, and can indicate the comprehensiveness the genome annotation. Replicates that are obvious outliers may have sequencing/technical issues causing false detection of splice junctions.  
\par
In general, abnormalities in the splice junction rates are generally a symptom of larger issues which will often be picked up by other metrics. See Section \ref{subsec:plotDesc-13}.
\end{framed}%
\lthtmlfigureZ
\lthtmlcheckvsize\clearpage}

\stepcounter{subsubsection}
{\newpage\clearpage
\lthtmlfigureA{framed1280}%
\begin{framed}
% latex2html id marker 1280
 \small
\emph{What it means and what to look for:} This plot is useful for identifying mapping and/or annotation issues, and can indicate the comprehensiveness the genome annotation. Replicates that are obvious outliers may have sequencing/technical issues causing false detection of splice junctions.  
\par
In general, abnormalities in the splice junction rates are generally a symptom of larger issues which will often be picked up by other metrics. See Section \ref{subsec:plotDesc-13}.
\end{framed}%
\lthtmlfigureZ
\lthtmlcheckvsize\clearpage}

\stepcounter{subsubsection}
{\newpage\clearpage
\lthtmlfigureA{framed1299}%
\begin{framed} \small
\emph{What it means and what to look for:} This plot can indicate the efficiency of the strand-specific selection protocol, and reveal variations in such efficiency. They can also be used to determine the "strandedness rule", which is required by many downstream analysis tools.
\end{framed}%
\lthtmlfigureZ
\lthtmlcheckvsize\clearpage}

\stepcounter{subsubsection}
{\newpage\clearpage
\lthtmlfigureA{framed1319}%
\begin{framed} \small
\emph{What it means and what to look for:} Presence of outliers in the mapping rate statistics may be an indicator of large sample-prep, library-prep, or sequencer errors.
\end{framed}%
\lthtmlfigureZ
\lthtmlcheckvsize\clearpage}

\stepcounter{subsubsection}
{\newpage\clearpage
\lthtmlfigureA{framed1338}%
\begin{framed} \small
\emph{What it means and what to look for:} Presence of outliers in these plots may point to variable inefficiency in a ribosomal/mitochondrial depletion protocol. In most datasets the Y chromosome counts can be used to determine sample sex (but not in this case, since the Y chromosome is not included in the example dataset's genome assembly). The raw metrics generated by QoRTs can also be used to generate counts for the ERCC spike-ins or similar.
\end{framed}%
\lthtmlfigureZ
\lthtmlcheckvsize\clearpage}

\stepcounter{subsubsection}
{\newpage\clearpage
\lthtmlfigureA{framed1357}%
\begin{framed} \small
\emph{What it means and what to look for:} These normalization factors can be used for a number of downstream analyses, including the generation of summary browser tracks.
\end{framed}%
\lthtmlfigureZ
\lthtmlcheckvsize\clearpage}

\stepcounter{subsubsection}
{\newpage\clearpage
\lthtmlfigureA{framed1376}%
\begin{framed} \small
\emph{What it means and what to look for:} Large variations in these ratios can indicate large-scale differences between the samples.
\end{framed}%
\lthtmlfigureZ
\lthtmlcheckvsize\clearpage}

\stepcounter{subsubsection}
{\newpage\clearpage
\lthtmlfigureA{framed1393}%
\begin{framed} \small
\emph{What it means and what to look for:} This can be used to assess the occurrance rates of a number of failure modes. 
\end{framed}%
\lthtmlfigureZ
\lthtmlcheckvsize\clearpage}

\stepcounter{subsubsection}
{\newpage\clearpage
\lthtmlfigureA{framed1410}%
\begin{framed} \small
\emph{What it means and what to look for:} With certain sample/library prep protocols there may be substantial risk of irregularities in the quantity of certain types of RNA.
For example, flawed or insufficient ribo depletion may result in large quantities of rRNA.
\end{framed}%
\lthtmlfigureZ
\lthtmlcheckvsize\clearpage}

\stepcounter{subsubsection}
\stepcounter{subsubsection}
{\newpage\clearpage
\lthtmlfigureA{framed1454}%
\begin{framed} \small
\emph{What it means and what to look for:} Excessive trimming may be indicative of other problems. It may also cause apparent differentials elsewhere (for example, it may reduce the rate at which splice junctions are observed, since long reads are needed to map across splice junctions).
\end{framed}%
\lthtmlfigureZ
\lthtmlcheckvsize\clearpage}

\stepcounter{subsubsection}
{\newpage\clearpage
\lthtmlfigureA{framed1470}%
\begin{framed} \small
\emph{What it means and what to look for:} This plot is mostly intended for benchmarking.
\end{framed}%
\lthtmlfigureZ
\lthtmlcheckvsize\clearpage}

\stepcounter{subsubsection}
{\newpage\clearpage
\lthtmlfigureA{framed1486}%
\begin{framed} \small
\emph{What it means and what to look for:} This plot may sometimes reveal the underlying causes behind discrepancies in the overlap-mismatch rates. Large differences in the amount of overlap may cause apparent differences in the overlap mismatch rates.
\end{framed}%
\lthtmlfigureZ
\lthtmlcheckvsize\clearpage}

\stepcounter{subsubsection}
{\newpage\clearpage
\lthtmlfigureA{framed1504}%
\begin{framed} \small
\emph{What it means and what to look for:} Excessive overlap errors may be indicative of excessive sequencing error.
\end{framed}%
\lthtmlfigureZ
\lthtmlcheckvsize\clearpage}

\stepcounter{subsubsection}
\stepcounter{subsubsection}
{\newpage\clearpage
\lthtmlfigureA{framed1538}%
\begin{framed} \small
\emph{What it means and what to look for:} Excessive overlap errors may be indicative of excessive sequencing error. However, if the excessive errors are all on base-pairs that are marked with low quality, you may be able to filter out such errors and prevent them from impacting downstream analyses.
\end{framed}%
\lthtmlfigureZ
\lthtmlcheckvsize\clearpage}

\stepcounter{subsubsection}
\stepcounter{subsubsection}
{\newpage\clearpage
\lthtmlfigureA{framed1572}%
\begin{framed} \small
\emph{What it means and what to look for:} Sometimes excessive overlap mismatches are disproportionately represented by a specific base-pair swap.
\end{framed}%
\lthtmlfigureZ
\lthtmlcheckvsize\clearpage}

\stepcounter{subsubsection}
\stepcounter{subsubsection}
{\newpage\clearpage
\lthtmlfigureA{framed1611}%
\begin{framed} \small
\emph{What it means and what to look for:} Excessive reference mismatches may be caused by a wide range of different factors. This includes sequencing error (in which case we would expect to also see excessive overlap mismatches), population inhomogeneity, RNA/DNA degradation, library prep issues, or even the activity (or lack thereof) of DNA repair genes.
\end{framed}%
\lthtmlfigureZ
\lthtmlcheckvsize\clearpage}

\stepcounter{subsubsection}
{\newpage\clearpage
\lthtmlfigureA{framed1628}%
\begin{framed} \small
\emph{What it means and what to look for:} In some cases, excessive reference mismatches may be marked with lower PHRED scores. If this is the case, it may be possible to filter for this effect or otherwise reduce its impact. If there are excessive high-quality mismatches, this may be more of a problem.
\end{framed}%
\lthtmlfigureZ
\lthtmlcheckvsize\clearpage}

\stepcounter{subsubsection}
{\newpage\clearpage
\lthtmlfigureA{framed1645}%
\begin{framed} \small
\emph{What it means and what to look for:} In many cases, excessive reference mismatches may be disproportionately represented by a specific base-pair swap.
\end{framed}%
\lthtmlfigureZ
\lthtmlcheckvsize\clearpage}

\stepcounter{subsubsection}
\stepcounter{section}
\stepcounter{subsection}
\stepcounter{subsection}
\stepcounter{section}
\stepcounter{subsection}
\stepcounter{subsection}
\stepcounter{subsection}
\stepcounter{subsubsection}
\stepcounter{subsubsection}
\stepcounter{subsubsection}
\stepcounter{subsubsection}
\stepcounter{subsubsection}
\stepcounter{subsection}
\stepcounter{subsubsection}
\stepcounter{section}
\stepcounter{section}
{\newpage\clearpage
\lthtmlfigureA{framed1805}%
\begin{framed} \begin{verbatim}

java -jar /path/to/jarfile/QoRTs.jar "?" samjdkinfo\end{verbatim}
  \end{framed}%
\lthtmlfigureZ
\lthtmlcheckvsize\clearpage}


\end{document}
